### PR TITLE
Add the ability for providers to trigger upgrades based on spec changes

### PR DESCRIPTION
**Pull request by g-gaston**
*Description of changes:*
Sometimes, the spec doesn't change and there isn't a new eks-d release in the bundle. However, it's possible that some of the components in the bundle used by a provider have changed, and a new "upgrade" (just reapply the templates) is needed.

This PR adds a new method to providers to allow them to trigger upgrades even when cluster manager doesn't detect a change.
I think we should eventually move the provider specific code we have in cluster manager to these methods (that's why I added the `Context` as an argument), but I'll leave that refactor for a separate PR.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/547#" title="Author self-approved">g-gaston</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [g-gaston]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by abhay-krishna** [comment]
@g-gaston: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-attribution-files-presubmit | 687c39a49783eaf7a19390691648a42b7ebb2d2b | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/547/eks-anywhere-attribution-files-presubmit/1455547055937163264) | `/test eks-anywhere-attribution-files-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=547). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@g-gaston: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-attribution-files-presubmit | 687c39a49783eaf7a19390691648a42b7ebb2d2b | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/547/eks-anywhere-attribution-files-presubmit/1455547055937163264) | `/test eks-anywhere-attribution-files-presubmit`
eks-anywhere-mocks-presubmits | 687c39a49783eaf7a19390691648a42b7ebb2d2b | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/547/eks-anywhere-mocks-presubmits/1455547056021049344) | `/test eks-anywhere-mocks-presubmits`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=547). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@g-gaston: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-attribution-files-presubmit | 687c39a49783eaf7a19390691648a42b7ebb2d2b | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/547/eks-anywhere-attribution-files-presubmit/1455547055937163264) | `/test eks-anywhere-attribution-files-presubmit`
eks-anywhere-mocks-presubmits | 687c39a49783eaf7a19390691648a42b7ebb2d2b | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/547/eks-anywhere-mocks-presubmits/1455547056021049344) | `/test eks-anywhere-mocks-presubmits`
eks-anywhere-e2e-presubmit | 687c39a49783eaf7a19390691648a42b7ebb2d2b | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/547/eks-anywhere-e2e-presubmit/1455547055979106304) | `/test eks-anywhere-e2e-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=547). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Review by mdsgabriel — APPROVED** [review]

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/547#" title="Author self-approved">g-gaston</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/547#pullrequestreview-795617092" title="Approved">mdsgabriel</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [g-gaston,mdsgabriel]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Review by jiayiwang7 — APPROVED** [review]

**Comment by abhay-krishna** [comment]
New changes are detected. LGTM label has been removed.

**Pull request merged**
